File: erpnext/crm/chat_crypto.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt
"""Key management for secret (end-to-end encrypted) chats.

The server is deliberately a dumb key directory: it stores public keys, the user's
private key *already wrapped* by a key derived from their passphrase in the browser,
and per-thread keys wrapped for each participant. It never sees a passphrase, an
unwrapped private key, or a plaintext message body — losing the passphrase means the
history is gone, by design (see plans/secret chats).

Sealing (`seal_for_users`) is the one place the server touches ciphertext: WhatsApp
messages arrive in the clear from Meta and are encrypted *to* the managers' public keys
before they are written. That needs public keys only — no private key ever lives here.
"""

import base64
import json
import os

import frappe
from frappe import _
from frappe.utils import now

ENC_VERSION = 1
SEAL_ALG = "ECDH-ES-P256-A256GCM"


# ---------------------------------------------------------------------------
# Enrolment
# ---------------------------------------------------------------------------


def _my_key_doc(user=None):
	user = user or frappe.session.user
	name = frappe.db.exists("Chat Encryption Key", {"user": user})
	return frappe.get_doc("Chat Encryption Key", name) if name else None


def is_enrolled(user):
	return bool(frappe.db.exists("Chat Encryption Key", {"user": user}))


@frappe.whitelist()
def enroll(public_key, wrapped_private_key, kdf_salt, signing_public_key=None, kdf_iterations=600000):
	"""Register the current user's key material. Re-enrolling replaces the identity and
	orphans every existing thread key — the old history becomes unreadable, so the UI
	must warn before calling this a second time."""
	me = frappe.session.user
	if me == "Guest":
		frappe.throw(_("Not permitted"), frappe.PermissionError)

	doc = _my_key_doc(me)
	if not doc:
		doc = frappe.new_doc("Chat Encryption Key")
		doc.user = me
	else:
		doc.key_version = (doc.key_version or 1) + 1
		doc.devices = []

	doc.public_key = public_key
	doc.signing_public_key = signing_public_key
	doc.wrapped_private_key = wrapped_private_key
	doc.kdf_salt = kdf_salt
	doc.kdf_iterations = int(kdf_iterations or 600000)
	doc.enrolled_on = now()
	doc.save(ignore_permissions=True)
	return get_my_key()


@frappe.whitelist()
def get_my_key():
	"""Everything the browser needs to unlock: the wrapped private key, its KDF
	parameters, and the registered biometric devices."""
	doc = _my_key_doc()
	if not doc:
		return None
	return {
		"user": doc.user,
		"public_key": doc.public_key,
		"signing_public_key": doc.signing_public_key,
		"wrapped_private_key": doc.wrapped_private_key,
		"kdf_salt": doc.kdf_salt,
		"kdf_iterations": doc.kdf_iterations or 600000,
		"key_version": doc.key_version or 1,
		"enrolled_on": str(doc.enrolled_on) if doc.enrolled_on else None,
		"devices": [
			{
				"name": d.name,
				"label": d.label,
				"credential_id": d.credential_id,
				"prf_salt": d.prf_salt,
				"wrapped_private_key": d.wrapped_private_key,
				"last_used_on": str(d.last_used_on) if d.last_used_on else None,
			}
			for d in doc.devices
		],
	}


@frappe.whitelist()
def change_passphrase(wrapped_private_key, kdf_salt, kdf_iterations=600000):
	"""Store the private key re-wrapped under a new passphrase. The identity keypair is
	unchanged, so every existing thread key keeps working."""
	doc = _my_key_doc()
	if not doc:
		frappe.throw(_("Secret chats are not enabled for your account"))
	doc.wrapped_private_key = wrapped_private_key
	doc.kdf_salt = kdf_salt
	doc.kdf_iterations = int(kdf_iterations or 600000)
	doc.save(ignore_permissions=True)
	return {"ok": True}


@frappe.whitelist()
def get_public_keys(users):
	"""Public keys of the given users — the input for wrapping a thread key. Users who
	have not enabled secret chats are simply absent from the result."""
	if isinstance(users, str):
		users = frappe.parse_json(users)
	if not users:
		return {}
	rows = frappe.get_all(
		"Chat Encryption Key",
		filters={"user": ["in", list(users)]},
		fields=["user", "public_key", "signing_public_key", "key_version"],
	)
	return {r.user: r for r in rows}


@frappe.whitelist()
def get_enrolled_users():
	"""Users who can take part in a secret chat."""
	return frappe.get_all("Chat Encryption Key", pluck="user")


# ---------------------------------------------------------------------------
# Biometric devices (WebAuthn PRF)
# ---------------------------------------------------------------------------


@frappe.whitelist()
def register_device(credential_id, prf_salt, wrapped_private_key, label=None):
	"""Add a per-device copy of the private key, wrapped with the secret the device's
	authenticator derives (Touch ID / Windows Hello). The passphrase stays the root."""
	doc = _my_key_doc()
	if not doc:
		frappe.throw(_("Secret chats are not enabled for your account"))

	for row in doc.devices:
		if row.credential_id == credential_id:
			row.prf_salt = prf_salt
			row.wrapped_private_key = wrapped_private_key
			row.label = label or row.label
			break
	else:
		doc.append(
			"devices",
			{
				"label": label or _("This device"),
				"credential_id": credential_id,
				"prf_salt": prf_salt,
				"wrapped_private_key": wrapped_private_key,
			},
		)
	doc.save(ignore_permissions=True)
	return get_my_key()


@frappe.whitelist()
def touch_device(credential_id):
	"""Record a successful biometric unlock (shown in the device list)."""
	doc = _my_key_doc()
	if not doc:
		return
	for row in doc.devices:
		if row.credential_id == credential_id:
			frappe.db.set_value(
				"Chat Device Key", row.name, "last_used_on", now(), update_modified=False
			)
			break


@frappe.whitelist()
def revoke_device(name):
	"""Drop a device's wrapped key — that device can no longer unlock without the
	passphrase."""
	doc = _my_key_doc()
	if not doc:
		return
	doc.devices = [d for d in doc.devices if d.name != name]
	doc.save(ignore_permissions=True)
	return get_my_key()


# ---------------------------------------------------------------------------
# Thread keys
# ---------------------------------------------------------------------------


def _assert_thread_member(thread, thread_doctype):
	"""Only a participant may read or hand out that thread's wrapped keys."""
	if thread_doctype == "WhatsApp Chat":
		from erpnext.crm.page.whatsapp_chat.whatsapp_chat import _require_wa_access

		_require_wa_access()
		return

	from erpnext.crm.page.employee_chat.employee_chat import _require_participant

	return _require_participant(thread)


@frappe.whitelist()
def get_thread_key(thread, thread_doctype="Chat Thread"):
	"""The current user's wrapped copy of a thread key."""
	_assert_thread_member(thread, thread_doctype)
	row = frappe.db.get_value(
		"Chat Thread Key",
		{"thread": thread, "user": frappe.session.user},
		["name", "wrapped_thread_key", "ephemeral_public_key", "alg"],
		as_dict=True,
	)
	return row


@frappe.whitelist()
def get_thread_keys(threads, thread_doctype="Chat Thread"):
	"""Batch form of `get_thread_key` — used when the chat list opens."""
	if isinstance(threads, str):
		threads = frappe.parse_json(threads)
	if not threads:
		return {}
	rows = frappe.get_all(
		"Chat Thread Key",
		filters={"thread": ["in", list(threads)], "user": frappe.session.user},
		fields=["thread", "wrapped_thread_key", "ephemeral_public_key", "alg"],
	)
	return {r.thread: r for r in rows}


def store_thread_key(thread, user, wrapped_thread_key, ephemeral_public_key, thread_doctype="Chat Thread", granted_by=None):
	"""Upsert one participant's wrapped copy of a thread key."""
	existing = frappe.db.exists("Chat Thread Key", {"thread": thread, "user": user})
	doc = (
		frappe.get_doc("Chat Thread Key", existing)
		if existing
		else frappe.new_doc("Chat Thread Key")
	)
	doc.thread = thread
	doc.thread_doctype = thread_doctype
	doc.user = user
	doc.wrapped_thread_key = wrapped_thread_key
	doc.ephemeral_public_key = ephemeral_public_key
	doc.alg = SEAL_ALG
	doc.granted_by = granted_by or frappe.session.user
	doc.save(ignore_permissions=True)
	return doc.name


@frappe.whitelist()
def grant_thread_key(thread, keys, thread_doctype="Chat Thread"):
	"""Hand out a thread key to participants. `keys` is a JSON list of
	{user, wrapped_thread_key, ephemeral_public_key} produced in the granter's browser —
	the plaintext thread key never reaches the server."""
	_assert_thread_member(thread, thread_doctype)
	if isinstance(keys, str):
		keys = frappe.parse_json(keys)

	granted = []
	for k in keys or []:
		user = k.get("user")
		if not user or not k.get("wrapped_thread_key"):
			continue
		store_thread_key(
			thread,
			user,
			k["wrapped_thread_key"],
			k.get("ephemeral_public_key"),
			thread_doctype=thread_doctype,
		)
		granted.append(user)
	return {"granted": granted}


def drop_thread_keys(thread, user):
	"""Revoke a user's access to a thread (e.g. removed from a group). Past messages they
	already read are of course beyond recall."""
	for name in frappe.get_all(
		"Chat Thread Key", filters={"thread": thread, "user": user}, pluck="name"
	):
		frappe.delete_doc("Chat Thread Key", name, ignore_permissions=True, force=True)


# ---------------------------------------------------------------------------
# Server-side sealing (WhatsApp ingest)
# ---------------------------------------------------------------------------


def _b64(raw):
	return base64.b64encode(raw).decode()


def _unb64(text):
	return base64.b64decode(text)


def seal_for_users(payload, users):
	"""Encrypt `payload` (a JSON-serialisable dict) so that only `users` can read it.

	A fresh AES-256-GCM content key is generated, and wrapped for every recipient with
	ECDH-ES against their public key — which needs the *public* half only, so the server
	stays unable to decrypt what it just wrote. Returns None when nobody can receive it
	(no enrolled recipient), so callers can decide to drop or keep the message.
	"""
	from cryptography.hazmat.primitives import hashes, serialization
	from cryptography.hazmat.primitives.asymmetric import ec
	from cryptography.hazmat.primitives.ciphers.aead import AESGCM
	from cryptography.hazmat.primitives.kdf.hkdf import HKDF

	recipients = get_public_keys(list(users or []))
	if not recipients:
		return None

	content_key = AESGCM.generate_key(bit_length=256)
	iv = os.urandom(12)
	ciphertext = AESGCM(content_key).encrypt(iv, json.dumps(payload).encode(), None)

	wrapped = []
	for user, row in recipients.items():
		peer = serialization.load_der_public_key(_unb64(row.public_key))
		ephemeral = ec.generate_private_key(ec.SECP256R1())
		shared = ephemeral.exchange(ec.ECDH(), peer)
		kek = HKDF(
			algorithm=hashes.SHA256(), length=32, salt=None, info=SEAL_ALG.encode()
		).derive(shared)
		wrap_iv = os.urandom(12)
		wrapped_key = AESGCM(kek).encrypt(wrap_iv, content_key, None)
		wrapped.append(
			{
				"user": user,
				# iv || ciphertext, so the browser can split it without extra fields
				"wrapped_thread_key": _b64(wrap_iv + wrapped_key),
				"ephemeral_public_key": _b64(
					ephemeral.public_key().public_bytes(
						serialization.Encoding.DER,
						serialization.PublicFormat.SubjectPublicKeyInfo,
					)
				),
			}
		)

	return {
		"ciphertext": _b64(ciphertext),
		"iv": _b64(iv),
		"enc_version": ENC_VERSION,
		"keys": wrapped,
	}

File: erpnext/public/js/chat_crypto.js
// End-to-end encryption for secret chats (Employee Chat + WhatsApp Chat Center).
//
// Everything secret happens here, in the browser. The server stores public keys, the
// private key *already wrapped* with a key derived from the user's passphrase, and the
// per-thread keys wrapped for each participant. It never receives a passphrase, an
// unwrapped private key, or a plaintext body — so a database dump is worthless, which is
// the whole point of the feature.
//
// Consequence to keep in mind while editing: there is no recovery path. Lose the
// passphrase and every registered device, and the history is gone for good.

frappe.provide("erpnext.chat_crypto");

const CRYPTO_API = "erpnext.crm.chat_crypto.";
const KDF_ITERATIONS = 600000;
// Must match SEAL_ALG in erpnext/crm/chat_crypto.py — it is the HKDF info string, so a
// mismatch silently produces a different key and nothing decrypts.
const WRAP_INFO = "ECDH-ES-P256-A256GCM";
const EC_PARAMS = { name: "ECDH", namedCurve: "P-256" };
const SIGN_PARAMS = { name: "ECDSA", namedCurve: "P-256" };

// Unlocked key material. Module-scoped on purpose: never localStorage, never
// sessionStorage, never on `window` — it dies with the tab.
let IDENTITY = null; // {ecdh: CryptoKeyPair-private, ecdsa: private, public_key: b64}
let MY_KEY = null; // cached Chat Encryption Key record
const THREAD_KEYS = {}; // thread -> CryptoKey (AES-GCM)
const subtle = () => window.crypto && window.crypto.subtle;

// --- encoding helpers ------------------------------------------------------

function to_b64(buf) {
	const bytes = new Uint8Array(buf);
	let s = "";
	for (let i = 0; i < bytes.length; i++) s += String.fromCharCode(bytes[i]);
	return btoa(s);
}

function from_b64(text) {
	const s = atob(text);
	const bytes = new Uint8Array(s.length);
	for (let i = 0; i < s.length; i++) bytes[i] = s.charCodeAt(i);
	return bytes;
}

function random_bytes(n) {
	return window.crypto.getRandomValues(new Uint8Array(n));
}

// --- key derivation --------------------------------------------------------

async function derive_kek(passphrase, salt, iterations) {
	const base = await subtle().importKey(
		"raw",
		new TextEncoder().encode(passphrase),
		"PBKDF2",
		false,
		["deriveKey"]
	);
	return subtle().deriveKey(
		{ name: "PBKDF2", salt, iterations: iterations || KDF_ITERATIONS, hash: "SHA-256" },
		base,
		{ name: "AES-GCM", length: 256 },
		false,
		["encrypt", "decrypt"]
	);
}

// The two private keys travel as one blob so a single unlock restores both.
async function wrap_identity(kek, ecdh_private, ecdsa_private) {
	const payload = JSON.stringify({
		ecdh: to_b64(await subtle().exportKey("pkcs8", ecdh_private)),
		ecdsa: ecdsa_private ? to_b64(await subtle().exportKey("pkcs8", ecdsa_private)) : null,
	});
	const iv = random_bytes(12);
	const ct = await subtle().encrypt(
		{ name: "AES-GCM", iv },
		kek,
		new TextEncoder().encode(payload)
	);
	return to_b64(new Uint8Array([...iv, ...new Uint8Array(ct)]));
}

async function unwrap_identity(kek, wrapped) {
	const raw = from_b64(wrapped);
	const iv = raw.slice(0, 12);
	const plain = await subtle().decrypt({ name: "AES-GCM", iv }, kek, raw.slice(12));
	const blob = JSON.parse(new TextDecoder().decode(plain));
	const ecdh = await subtle().importKey("pkcs8", from_b64(blob.ecdh), EC_PARAMS, true, [
		"deriveBits",
	]);
	const ecdsa = blob.ecdsa
		? await subtle().importKey("pkcs8", from_b64(blob.ecdsa), SIGN_PARAMS, true, ["sign"])
		: null;
	return { ecdh, ecdsa };
}

// ECDH-ES: one ephemeral keypair per wrap, HKDF over the shared secret. Mirrors
// seal_for_users() in chat_crypto.py so server-sealed WhatsApp messages open here too.
async function derive_wrap_key(private_key, peer_public_key) {
	const shared = await subtle().deriveBits(
		{ name: "ECDH", public: peer_public_key },
		private_key,
		256
	);
	const hkdf = await subtle().importKey("raw", shared, "HKDF", false, ["deriveKey"]);
	return subtle().deriveKey(
		{
			name: "HKDF",
			hash: "SHA-256",
			salt: new Uint8Array(0),
			info: new TextEncoder().encode(WRAP_INFO),
		},
		hkdf,
		{ name: "AES-GCM", length: 256 },
		false,
		["encrypt", "decrypt"]
	);
}

async function import_public(b64) {
	return subtle().importKey("spki", from_b64(b64), EC_PARAMS, true, []);
}

// --- WebAuthn PRF (biometric unlock) ---------------------------------------
//
// The passkey is not used to authenticate against the server — it is used purely as a
// key vault: the PRF extension returns a stable 32-byte secret that never leaves the
// device, and we wrap a second copy of the private key with it. That is why a locally
// generated challenge is fine here; there is nothing for the server to verify.

async function prf_secret(credential_id, salt_b64) {
	const assertion = await navigator.credentials.get({
		publicKey: {
			challenge: random_bytes(32),
			allowCredentials: credential_id
				? [{ id: from_b64(credential_id), type: "public-key" }]
				: [],
			userVerification: "required",
			timeout: 60000,
			extensions: { prf: { eval: { first: from_b64(salt_b64) } } },
		},
	});
	if (!assertion) throw new Error("no assertion");
	const results = assertion.getClientExtensionResults();
	if (!results || !results.prf || !results.prf.results || !results.prf.results.first) {
		throw new Error("prf-unsupported");
	}
	return { secret: results.prf.results.first, credential_id: to_b64(assertion.rawId) };
}

async function kek_from_prf(secret) {
	const hkdf = await subtle().importKey("raw", secret, "HKDF", false, ["deriveKey"]);
	return subtle().deriveKey(
		{
			name: "HKDF",
			hash: "SHA-256",
			salt: new Uint8Array(0),
			info: new TextEncoder().encode("chat-device-kek"),
		},
		hkdf,
		{ name: "AES-GCM", length: 256 },
		false,
		["encrypt", "decrypt"]
	);
}

erpnext.chat_crypto = {
	// --- state -------------------------------------------------------------

	is_unlocked() {
		return !!IDENTITY;
	},

	async is_enrolled() {
		return !!(await this.my_key());
	},

	async my_key(force) {
		if (MY_KEY && !force) return MY_KEY;
		MY_KEY = await frappe.xcall(CRYPTO_API + "get_my_key");
		return MY_KEY;
	},

	lock() {
		IDENTITY = null;
		for (const k of Object.keys(THREAD_KEYS)) delete THREAD_KEYS[k];
	},

	supported() {
		return !!subtle() && !!window.isSecureContext;
	},

	supports_biometric() {
		return !!(window.PublicKeyCredential && navigator.credentials && window.isSecureContext);
	},

	has_biometric() {
		return !!(MY_KEY && (MY_KEY.devices || []).length);
	},

	// --- enrolment ---------------------------------------------------------

	async enroll(passphrase) {
		if (!this.supported()) frappe.throw(__("This browser cannot do encrypted chats"));
		const salt = random_bytes(16);
		const kek = await derive_kek(passphrase, salt, KDF_ITERATIONS);

		const ecdh = await subtle().generateKey(EC_PARAMS, true, ["deriveBits"]);
		const ecdsa = await subtle().generateKey(SIGN_PARAMS, true, ["sign", "verify"]);

		const payload = {
			public_key: to_b64(await subtle().exportKey("spki", ecdh.publicKey)),
			signing_public_key: to_b64(await subtle().exportKey("spki", ecdsa.publicKey)),
			wrapped_private_key: await wrap_identity(kek, ecdh.privateKey, ecdsa.privateKey),
			kdf_salt: to_b64(salt),
			kdf_iterations: KDF_ITERATIONS,
		};
		MY_KEY = await frappe.xcall(CRYPTO_API + "enroll", payload);
		IDENTITY = { ecdh: ecdh.privateKey, ecdsa: ecdsa.privateKey };
		return MY_KEY;
	},

	async unlock(passphrase) {
		const key = await this.my_key(true);
		if (!key) frappe.throw(__("Secret chats are not enabled for your account"));
		const kek = await derive_kek(
			passphrase,
			from_b64(key.kdf_salt),
			key.kdf_iterations || KDF_ITERATIONS
		);
		try {
			IDENTITY = await unwrap_identity(kek, key.wrapped_private_key);
		} catch (e) {
			throw new Error("bad-passphrase");
		}
		return true;
	},

	async unlock_with_biometric() {
		const key = await this.my_key();
		const device = (key && key.devices && key.devices[0]) || null;
		if (!device) throw new Error("no-device");
		const { secret, credential_id } = await prf_secret(device.credential_id, device.prf_salt);
		const kek = await kek_from_prf(secret);
		IDENTITY = await unwrap_identity(kek, device.wrapped_private_key);
		frappe.xcall(CRYPTO_API + "touch_device", { credential_id }).catch(() => {});
		return true;
	},

	// Registering needs an unlocked identity: we re-wrap the private key with the
	// device secret and hand only the wrapped blob to the server.
	async register_biometric(label) {
		if (!IDENTITY) frappe.throw(__("Unlock secret chats first"));
		if (!this.supports_biometric()) frappe.throw(__("This device has no biometric support"));

		const user = frappe.session.user;
		const created = await navigator.credentials.create({
			publicKey: {
				rp: { id: window.location.hostname, name: "ERPNext" },
				user: {
					id: new TextEncoder().encode(user),
					name: user,
					displayName: frappe.session.user_fullname || user,
				},
				challenge: random_bytes(32),
				pubKeyCredParams: [
					{ type: "public-key", alg: -7 },
					{ type: "public-key", alg: -257 },
				],
				authenticatorSelection: {
					authenticatorAttachment: "platform",
					userVerification: "required",
					residentKey: "preferred",
				},
				timeout: 60000,
				extensions: { prf: {} },
			},
		});
		if (!created) throw new Error("no-credential");

		// The PRF output is only available from an assertion, so ask for one right away.
		const salt = random_bytes(32);
		const credential_id = to_b64(created.rawId);
		const { secret } = await prf_secret(credential_id, to_b64(salt));
		const kek = await kek_from_prf(secret);

		MY_KEY = await frappe.xcall(CRYPTO_API + "register_device", {
			credential_id,
			prf_salt: to_b64(salt),
			wrapped_private_key: await wrap_identity(kek, IDENTITY.ecdh, IDENTITY.ecdsa),
			label: label || __("This device"),
		});
		return MY_KEY;
	},

	async revoke_biometric(name) {
		MY_KEY = await frappe.xcall(CRYPTO_API + "revoke_device", { name });
		return MY_KEY;
	},

	async change_passphrase(new_passphrase) {
		if (!IDENTITY) frappe.throw(__("Unlock secret chats first"));
		const salt = random_bytes(16);
		const kek = await derive_kek(new_passphrase, salt, KDF_ITERATIONS);
		await frappe.xcall(CRYPTO_API + "change_passphrase", {
			wrapped_private_key: await wrap_identity(kek, IDENTITY.ecdh, IDENTITY.ecdsa),
			kdf_salt: to_b64(salt),
			kdf_iterations: KDF_ITERATIONS,
		});
		// Devices keep their own wrapping, but they were bound to the old passphrase's
		// trust decision — make the user re-register deliberately.
		await this.my_key(true);
		return true;
	},

	// --- thread keys -------------------------------------------------------

	// Fresh AES-256-GCM key plus a wrapped copy for every participant.
	async new_thread_key(users) {
		const key = await subtle().generateKey({ name: "AES-GCM", length: 256 }, true, [
			"encrypt",
			"decrypt",
		]);
		return { key, wrapped: await this.wrap_for_users(key, users) };
	},

	async wrap_for_users(key, users) {
		const pubs = await frappe.xcall(CRYPTO_API + "get_public_keys", {
			users: JSON.stringify(users),
		});
		const raw = await subtle().exportKey("raw", key);
		const out = [];
		for (const user of users) {
			const row = pubs[user];
			if (!row) continue;
			const ephemeral = await subtle().generateKey(EC_PARAMS, true, ["deriveBits"]);
			const wrap_key = await derive_wrap_key(
				ephemeral.privateKey,
				await import_public(row.public_key)
			);
			const iv = random_bytes(12);
			const ct = await subtle().encrypt({ name: "AES-GCM", iv }, wrap_key, raw);
			out.push({
				user,
				wrapped_thread_key: to_b64(new Uint8Array([...iv, ...new Uint8Array(ct)])),
				ephemeral_public_key: to_b64(
					await subtle().exportKey("spki", ephemeral.publicKey)
				),
			});
		}
		return out;
	},

	async thread_key(thread) {
		if (THREAD_KEYS[thread]) return THREAD_KEYS[thread];
		if (!IDENTITY) throw new Error("locked");

		const row = await frappe.xcall(CRYPTO_API + "get_thread_key", { thread });
		if (!row) throw new Error("no-key");
		const wrap_key = await derive_wrap_key(
			IDENTITY.ecdh,
			await import_public(row.ephemeral_public_key)
		);
		const blob = from_b64(row.wrapped_thread_key);
		const raw = await subtle().decrypt(
			{ name: "AES-GCM", iv: blob.slice(0, 12) },
			wrap_key,
			blob.slice(12)
		);
		THREAD_KEYS[thread] = await subtle().importKey("raw", raw, { name: "AES-GCM" }, true, [
			"encrypt",
			"decrypt",
		]);
		return THREAD_KEYS[thread];
	},

	// Share an existing thread key with someone who joins later.
	async grant_thread_key(thread, users) {
		const key = await this.thread_key(thread);
		const wrapped = await this.wrap_for_users(key, users);
		await frappe.xcall(CRYPTO_API + "grant_thread_key", {
			thread,
			keys: JSON.stringify(wrapped),
		});
		return wrapped;
	},

	// --- message payloads --------------------------------------------------

	async encrypt(thread, obj) {
		const key = await this.thread_key(thread);
		const iv = random_bytes(12);
		const ct = await subtle().encrypt(
			{ name: "AES-GCM", iv },
			key,
			new TextEncoder().encode(JSON.stringify(obj))
		);
		return { ciphertext: to_b64(ct), iv: to_b64(iv) };
	},

	async decrypt(thread, ciphertext, iv) {
		const key = await this.thread_key(thread);
		const plain = await subtle().decrypt(
			{ name: "AES-GCM", iv: from_b64(iv) },
			key,
			from_b64(ciphertext)
		);
		return JSON.parse(new TextDecoder().decode(plain));
	},

	// --- attachments -------------------------------------------------------
	//
	// Files get their own random key so the bytes can be re-shared without handing out
	// the thread key; that key rides inside the encrypted message payload.

	async encrypt_blob(blob) {
		const key = await subtle().generateKey({ name: "AES-GCM", length: 256 }, true, [
			"encrypt",
			"decrypt",
		]);
		const iv = random_bytes(12);
		const ct = await subtle().encrypt(
			{ name: "AES-GCM", iv },
			key,
			await blob.arrayBuffer()
		);
		return {
			blob: new Blob([ct], { type: "application/octet-stream" }),
			key: to_b64(await subtle().exportKey("raw", key)),
			iv: to_b64(iv),
		};
	},

	async decrypt_blob(array_buffer, key_b64, iv_b64, mime) {
		const key = await subtle().importKey(
			"raw",
			from_b64(key_b64),
			{ name: "AES-GCM" },
			false,
			["decrypt"]
		);
		const plain = await subtle().decrypt(
			{ name: "AES-GCM", iv: from_b64(iv_b64) },
			key,
			array_buffer
		);
		return new Blob([plain], { type: mime || "application/octet-stream" });
	},

	// --- UI ----------------------------------------------------------------

	// Resolves once the identity is unlocked, prompting only if needed. Tries biometrics
	// first when the device has a registered passkey.
	async ensure_unlocked() {
		if (IDENTITY) return true;
		const key = await this.my_key(true);
		if (!key) {
			await this.setup_dialog();
			return this.is_unlocked();
		}
		if (this.has_biometric() && this.supports_biometric()) {
			try {
				await this.unlock_with_biometric();
				return true;
			} catch (e) {
				// Cancelled, unsupported PRF, or a different device — fall back below.
			}
		}
		return this.unlock_dialog();
	},

	unlock_dialog() {
		return new Promise((resolve) => {
			const d = new frappe.ui.Dialog({
				title: __("Unlock secret chats"),
				fields: [
					{
						fieldtype: "Password",
						fieldname: "passphrase",
						label: __("Passphrase"),
						reqd: 1,
					},
					{
						fieldtype: "HTML",
						fieldname: "hint",
						options: `<p class="text-muted small">${__(
							"Your passphrase never leaves this browser. It cannot be reset — if you forget it, the history is lost."
						)}</p>`,
					},
				],
				primary_action_label: __("Unlock"),
				primary_action: async (values) => {
					try {
						await this.unlock(values.passphrase);
					} catch (e) {
						frappe.msgprint(__("Wrong passphrase"));
						return;
					}
					d.hide();
					if (this.supports_biometric() && !this.has_biometric()) {
						this.offer_biometric();
					}
					resolve(true);
				},
			});
			d.onhide = () => resolve(this.is_unlocked());
			d.show();
		});
	},

	setup_dialog() {
		return new Promise((resolve) => {
			const d = new frappe.ui.Dialog({
				title: __("Enable secret chats"),
				fields: [
					{
						fieldtype: "HTML",
						fieldname: "intro",
						options: `<p>${__(
							"Secret chats are encrypted in your browser. Nobody else — not a system manager, not someone with database access — can read them."
						)}</p><p class="text-danger">${__(
							"There is no recovery. If you forget this passphrase and lose your devices, the messages are gone forever."
						)}</p>`,
					},
					{
						fieldtype: "Password",
						fieldname: "passphrase",
						label: __("Passphrase"),
						reqd: 1,
					},
					{
						fieldtype: "Password",
						fieldname: "confirm",
						label: __("Repeat passphrase"),
						reqd: 1,
					},
				],
				primary_action_label: __("Enable"),
				primary_action: async (values) => {
					if (values.passphrase !== values.confirm) {
						frappe.msgprint(__("The passphrases do not match"));
						return;
					}
					if ((values.passphrase || "").length < 8) {
						frappe.msgprint(__("Use at least 8 characters"));
						return;
					}
					await this.enroll(values.passphrase);
					d.hide();
					frappe.show_alert({ message: __("Secret chats enabled"), indicator: "green" });
					if (this.supports_biometric()) this.offer_biometric();
					resolve(true);
				},
			});
			d.onhide = () => resolve(this.is_unlocked());
			d.show();
		});
	},

	offer_biometric() {
		frappe.confirm(
			__("Use biometrics on this device instead of typing the passphrase every time?"),
			async () => {
				try {
					await this.register_biometric();
					frappe.show_alert({
						message: __("Biometric unlock enabled"),
						indicator: "green",
					});
				} catch (e) {
					frappe.msgprint(__("This device does not support biometric unlock"));
				}
			}
		);
	},
};

// The key must not outlive the tab.
$(window).on("beforeunload", () => erpnext.chat_crypto.lock());
